scene/resources/surface_tool.cpp

```diff
@@ -1163,14 +1164,61 @@ void SurfaceTool::generate_tangents() {
 	SMikkTSpaceContext msc;
 	msc.m_pInterface = &mkif;
 
-	TangentGenerationContextUserData triangle_data;
-	triangle_data.vertices = &vertex_array;
+	LocalVector<Vertex> valid_vertex_array;
+	LocalVector<int> valid_index_array;
+	for (unsigned int j = 0; j < index_array.size(); j += 3) {
+		if (!Math::is_finite((real_t)index_array[j + 0]) || !Math::is_finite((real_t)index_array[j + 1]) || !Math::is_finite((real_t)index_array[j + 2])) {
+			WARN_PRINT_ONCE("Non-finite index detected. Skipping triangle.");
+			continue;
+		}
+		Vector3 &v0 = vertex_array[index_array[j + 0]].vertex;
+		Vector3 &v1 = vertex_array[index_array[j + 1]].vertex;
+		Vector3 &v2 = vertex_array[index_array[j + 2]].vertex;
+
+		if (!Math::is_finite(v0.length()) || !Math::is_finite(v1.length()) || !Math::is_finite(v2.length())) {
+			WARN_PRINT_ONCE("Non-finite vertex detected. Skipping triangle.");
+			continue;
+		}
+
+		Vector3 face_normal = vec3_cross(v0 - v2, v0 - v1);
+		float face_area = face_normal.length();
+		if (!Math::is_finite(face_area) || face_area == 0) {
+			WARN_PRINT_ONCE("Ignoring triangle with non-finite normal in tangent generation.");
+			continue;
+		}
+
+		valid_vertex_array.push_back(vertex_array[index_array[j + 0]]);
+		valid_vertex_array.push_back(vertex_array[index_array[j + 1]]);
+		valid_vertex_array.push_back(vertex_array[index_array[j + 2]]);
+
+		valid_index_array.push_back(index_array[j + 0]);
+		valid_index_array.push_back(index_array[j + 1]);
+		valid_index_array.push_back(index_array[j + 2]);
+	}
+
+	TangentGenerationContextUserData triangle_data{};
+	triangle_data.vertices = &valid_vertex_array;
+	triangle_data.indices = &valid_index_array;
+	msc.m_pUserData = &triangle_data;
```

Comment:
This seems suspicious as it might make the indices not match up with the original vertex indices when `mikktspace.c` eventually calls `m_setTSpace()`, resulting in mismatched tangents for vertices.

If the goal was to just optimize mikktspace, I think it should not change the behavior of the library itself, meaning that there shouldn't need to be any changes in user code to accommodate it (except of course for supplying the direct vertex pointers, which is required). So I'd recommend removing this logic and fixing whatever difference there is between the unoptimized and optimized within `mikktspace.c` itself. Ideally all the optimizations within that should be semantically equivalent to the original mikktspace.c code.